relentless/optimize/method.py

```diff
@@ -171,30 +183,139 @@ def optimize(self, objective):
-            for x in dvars:
-                x.value -= self.step_size*res.gradient(x)
-            converged = self.has_converged(res)
+        cur_res = objective.compute()
+        while not self.has_converged(cur_res) and iter_num < self.max_iter:
+
+            #steepest descent update
+            for x in dvars: #is there any way to make this into a single line?
+                x.value = cur_res.design_variables[x] - self.step_size*cur_res.gradient[x]
+            next_res = objective.compute()
+
+            #if line search, attempt backtracking in interval
+            if self.line_search is not None:
+                line_res = self.line_search.find(objective=objective, start=cur_res, end=next_res)
+                for x in dvars:
+                    x.value = line_res.design_variables[x]
+                next_res = line_res
+
+            #recycle next result
+            cur_res = next_res
+
             iter_num += 1
 
-        return converged
+        return self.has_converged(cur_res)
+
+    @property
+    def max_iter(self):
+        """int: The maximum number of optimization iterations allowed."""
+        return self._max_iter
+
+    @max_iter.setter
+    def max_iter(self, value):
+        if not isinstance(value, int):
+            raise TypeError('The maximum number of iterations must be an integer.')
+        if value < 1:
+            raise ValueError('The maximum number of iterations must be positive.')
+        self._max_iter = value
 
     @property
     def step_size(self):
-        """float: The step size hyperparameter (:math:`\alpha`)."""
+        r"""float: The step size hyperparameter (:math:`\alpha`)."""
         return self._step_size
 
     @step_size.setter
     def step_size(self, value):
         if value <= 0:
-            raise ValueError('The step size must be positive.')
+            raise ValueError('The specified step size must be positive.')
         self._step_size = value
 
+    @property
+    def line_search(self):
+        """:class:`LineSearch`: The line search object used to optimize the step size."""
+        return self._line_search
+
+    @line_search.setter
+    def line_search(self, value):
+        if value is not None and not isinstance(value, LineSearch):
+            raise TypeError('If defined, the line search parameter must be a LineSearch object.')
+        self._line_search = value
+
+#TODO: document
+class LineSearch:
+    """
+    """
+    def __init__(self, abs_tol, max_iter):
+        self.abs_tol = abs_tol
+        self.max_iter = max_iter
+
+    def find(self, objective, start, end):
+        """
+        """
+        ovars = {x: x.value for x in objective.design_variables()}
+
+        # compute normalized search direction
+        d = end.design_variables - start.design_variables
+        max_step = d.norm()
+        d /= max_step
+
+        # compute start and end target values
+        targets = np.array([d.dot(-start.gradient), d.dot(-end.gradient)])
```

Comment:
For a minor performance difference, but the minus sign outside of the dot. This avoids making a copy of the gradient array before dotting.
```suggestion
        targets = np.array([-d.dot(start.gradient), -d.dot(end.gradient)])
```